src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlDelegatedTransaction.cs

```diff
@@ -208,9 +211,22 @@ public byte[] Promote()
                 }
 
                 //Throw exception only if Transaction is still active and not yet aborted.
-                if (promoteException != null && Transaction.TransactionInformation.Status != TransactionStatus.Aborted)
+                if (promoteException != null)
                 {
-                    throw SQL.PromotionFailed(promoteException);
+                    try
+                    {
+                        // Safely access Transaction status - as it's possible Transaction is not in right state.
+                        if (Transaction?.TransactionInformation?.Status != TransactionStatus.Aborted)
+                        {
+                            throw SQL.PromotionFailed(promoteException);
+                        }
+                    }
+                    catch (TransactionException te)
+                    {
+                        SqlClientEventSource.Log.TryTraceEvent("SqlDelegatedTransaction.Promote | RES | CPOOL | Object Id {0}, Client Connection Id {1}, Transaction exception occurred: {2}.", ObjectID, usersConnection?.ClientConnectionId, te.Message);
+                        // Throw promote exception if transaction state is unknown.
+                        throw SQL.PromotionFailed(promoteException);
```

Comment:
Shouldn't it just rethrow?

Reply:
We want to throw the original "promoteException" instead of the captured transaction exception.